(fix): Condensation events to reconstruct contexts added to event stream

csmith49 opened the pull request:
- [ ] This change is worth documenting at https://docs.all-hands.dev/
- [ ] Include this change in the Release Notes. If checked, you **must** provide an **end-user friendly** description for your change below

**End-user friendly description of the problem this fixes or functionality that this introduces.**


---
**Give a summary of what the PR does, explaining any non-trivial design decisions.**

This PR updates the condenser interface to support a stateless implementation of a rolling condenser.

Instead of returning just a condensed history, this new interface allows a condenser to return a `View` of the history or a `Condensation` event. The condensation event can be added to the event stream, which is then one step closer to being the only source of state needed for the agent.

To implement this strategy, rolling condensers now define three abstract functions:
* `get_view`, which converts the sequence of events in the state to a view object by utilizing the information in the condensation events,
* `get_condensation`, which takes a view and produces a condensation event defining how future views should be constructed, and
* `should_condense`, a trigger which determines when the condensation should happen.

In addition to the condenser changes, this PR updates the CodeAct agent implementation to use the new interface and updates all condenser tests accordingly.

---
**Link of any specific issues this addresses.**

This directly solves the problem in #6706, but should also resolve #7175.

enyst commented:
Thank you for this! It seems to do exactly what I hit a problem with, in the [other PR](https://github.com/All-Hands-AI/OpenHands/pull/7311).

I started at the other end of this issue (put condensation events in the stream), and meant to keep it minimal, just to see what it looks like, but such refactoring needs _also_ something like this PR, in order to actually work. 😅

csmith49 commented:
> Thank you for this! It seems to do exactly what I hit a problem with, in the [other PR](https://github.com/All-Hands-AI/OpenHands/pull/7311).
> 
> I started at the other end of this issue (put condensation events in the stream), and meant to keep it minimal, just to see what it looks like, but such refactoring needs _also_ something like this PR, in order to actually work. 😅

I totally missed that PR! Had my head down digging into some metrics.

This approach turned out way less intensive than I expected. I've got the condensers doing the right thing, just now getting to the point of testing the solution with the agent controller in the mix. Also need to figure out how to test the failure case we were seeing...

enyst commented:
> I've got the condensers doing the right thing, just now getting to the point of testing the solution with the agent controller in the mix. Also need to figure out how to test the failure case we were seeing...

On the latter, maybe you want to see the 2 comments with debugging info. Idk if it's useful, they're literally like random notes though. The thing is, I got a great event stream, totally by accident, it was exactly a few tokens away from ContextWindowExceededError 😅. So on restore, if the first message is like 10-ish tokens, it got the error, and if it was less, it didn't.

https://github.com/All-Hands-AI/OpenHands/pull/7311#issuecomment-2734691216

I literally removed newer events from my stream and restored the first ones repeatedly. That said, I was looking at more things than condensation actually (some strange LLM behavior on microagents)

enyst commented:
> I haven't dug through your PR yet. Do you think our implementations can just be smashed together?

I am still to check your latest changes, I think we may want to sync at least on what kind of Action do we want here, it kinda matters for the rest IMHO. This [kind of thing](https://github.com/All-Hands-AI/OpenHands/pull/7311/files#diff-14ed67fced390d6f697f7034f83afe61d7f64d8b6b6ff04efa535446f73ee459R193) may help merging them (I meant to not change _how_ the condensers work), idk.

csmith49 commented:
> On the latter, maybe you want to see the 2 comments with debugging info. Idk if it's useful, they're literally like random notes though. The thing is, I got a great event stream, totally by accident, it was exactly a few tokens away from ContextWindowExceededError 😅. So on restore, if the first message is like 10-ish tokens, it got the error, and if it was less, it didn't.

I'll take a closer look, thanks. It'd be great if we could also get the condensation trigger behavior to be token-count aware, even if we're just approximating it. With enough of a safety factor we might be able to avoid context window errors entirely.

> I am still to check your latest changes, I think we may want to sync at least on what kind of Action do we want here, it kinda matters for the rest IMHO.

My first pass had different events for each kind of condenser, because they wanted slightly different bits of information. The current approach just has a single event with the union of all those features. I'd love to have some kind of "payload polymorphism" so that even with a single event there's still some type-safety, but the current event implementation makes that pretty fiddly. Might revisit that after a little event clean-up.

As it stands, I've got all existing tests working and have confirmed condensation happens appropriately without disrupting the agent flow. There's still some clean-up needed on documentation, and I want to add more tests, but I'm decently happy with where this PR is now.

enyst commented on openhands/events/action/agent.py:
```diff
@@ -126,3 +126,18 @@ def __str__(self) -> str:
+
+    forgotten_event_ids: list[int] = field(default_factory=list)
+    considered_event_ids: list[int] = field(default_factory=list)
+    summary: str | None = None
```
Just a quick question (I'm also checking now), I wonder if `considered_event_ids` is useful for the summarizing condenser and what does it represent, or is it only for others?

I wonder if we need _lists_ in general. In the alternative in 7311, I was looking at
```
    start_id: int = -1
    end_id: int = -1
    summary: str = ''
    action: str = ActionType.CONDENSE
```

That seems to be the minimum needed to store in the Event, because we always have `state.history` so we can compute the rest. The meaning would be the summary of events between `start_id` and `end_id`. Do we ever have non-consecutive events?

csmith49 replied:
> Just a quick question (I'm also checking now), I wonder if considered_event_ids is useful for the summarizing condenser and what does it represent, or is it only for others?

Only useful for folks staring at the event stream. The considered event IDs are all events in the view the condenser uses to compute the condensation event. Part of my goal was to capture all the relevant parts of the condenser's decision-making so that we can surface it elsewhere.

> Do we ever have non-consecutive events?

The forgetting condensers (LLM summarizing and amortized forgetting) always dump them in blocks. But the LLM attention condenser can pick any subset of events to drop, so the list of event IDs is necessary there.

But I think the list is good for interpretability outside the agent loop. Makes it easy to reconstruct the exact decision the condenser made without having to worry about assumptions like the fact that the trajectory we're analyzing is a one-to-one copy of the history the condenser saw.

enyst replied:
Ah, I see, that makes sense.

One thing that we may want to consider here, about taking data outside the agent loop, is that we also have a way of getting information for evaluations: we store the necessary data in `State`. I think we do that already for condensing, using `extra_data`? It just stored only some basic info until now, but it doesn't have to be limited to it.

As another data point, `State`'s `last_error` is really an "error for eval consideration" in the sense that, AFAIK, that attribute existed only _for evals scripts_ to get for sure the last error that the agent encountered before finishing the loop.

I'm not sure if I suggest to do it 🤔, please lets just take it as food for thought.

We know all too well there are significant advantages in storing info directly in the stream. One disadvantage with storing too much info in the stream, though, is that in time, it makes the streams large, so large that running scripts on eval trajectories has slowed to a crawl, even if running scripts that only need some data, not all data, it is still loading everything and parsing it from json. 😓 Actually, I should propose some simplification there - will do today

enyst replied:
(strangely related, but a side issue: https://github.com/All-Hands-AI/OpenHands/pull/6913 : Maybe the case for storing response_ids for MessageActions in the stream is perhaps stronger if we are even considering lists of events for non-application purposes. 6913 seems like it allows an easy fix on the tokens issue, since these response_ids are already in the stream for all other agent actions (part of toolcall metadata), and only missing for MessageActions. Can we give it another look? Thinking that even if we're wrong to do it, the risk is just an integer in the stream. I can make a quick extension of `ConversationMemory` built on it, to show us how it'd solve the problem - it would help my slow brain think a little more clearly - hopefully. 😅)

enyst replied:
I gave some thought to the question what kind of things we store in `events` vs what kind of things we store in `State`, and I think eval data belongs to `State`.

It may be difficult to define exactly what falls into each bucket, but as you can see in the linked PRs, there is redundant or useless information in the stream, it leads to issues with slowness, and it's not always obvious how to remove it later.

What do you think about storing in `State` what isn't necessary for the agent/application?

csmith49 replied:
Oh yeah, big evaluation data definitely shouldn't live in the event stream. Agreed it's a fine line determining what is/isn't necessary, and we probably don't want `State` to become a free-for-all.

Do we have an issue for the "too much stream info" problem? We might want to move this convo there.

enyst replied:
> Do we have an issue for the "too much stream info" problem?

We don't have one, we have 2 for now: the 2 issues linked below are about it : https://github.com/All-Hands-AI/OpenHands/pull/7353#issuecomment-2744193060

I think the part of this discussion that belongs here regards how exactly we architect AgentCondensationAction so that we both have what we need, and don't go too much on compounding that problem. :)

To see it better in code than in words, a (quick) alternative proposal on that part is here:
- https://github.com/All-Hands-AI/OpenHands/pull/7526

enyst commented:
Re:
> running scripts on eval trajectories has slowed to a crawl
> some simplification there

I'm pondering at least these two optimizations:
- https://github.com/All-Hands-AI/OpenHands/pull/7398
- https://github.com/All-Hands-AI/OpenHands/pull/7404

enyst commented on openhands/memory/condenser/impl/llm_attention_condenser.py:
```diff
@@ -43,15 +48,28 @@ def __init__(self, llm: LLM, max_size: int = 100, keep_first: int = 1):
+
+        for event in events:
+            if isinstance(event, CondensationAction):
+                forgotten_event_ids.extend(event.forgotten_event_ids)
```
For multiple CondensationAction, it seems that `forgotten_event_ids` would have duplicate ids?

csmith49 replied:
It's possible, but I don't think it's a problem. We just check for inclusion in the list so multiple entries won't hurt -- I'll convert to a set to make it more clear.

But there's some effort to ensure any IDs in `forgotten_event_ids` also appear in `considered_event_ids`. So if we don't ever consider an event in a view (because it's already been forgotten) it shouldn't show up again.

enyst replied:
What I was thinking here is that duplicates would be saved?, if there are duplicates.

csmith49 replied:
There shouldn't be any duplicates across condensation actions.

enyst replied:
I'll mark this as solved, it was, thank you!

neubig reviewed (approved):
I read through and this looks good to me! The logic is a bit complex, but maybe that can't be helped given what we want to do.

enyst commented:
There might be two areas where we can tweak a bit:
- we should save in the stream data necessary for the agent, but I hesitate whether we should now save data that is useful for external scripts/evals. I think usually we've been using the State pickle instead.
- some logic could maybe be moved to conversation_memory, in theory it's its job to figure out the events in history and make up from them the context for the LLM. Such refactoring may help to make the code a bit more obvious here.

I think the second could be in a follow-up, the first however seems important to me because the event stream is saved on user machines, and we care about compatibility afaik. We are able to load relatively old streams seamlessly (usually). Idk, this one would be more difficult to change later.

csmith49 commented:
> * we should save in the stream data necessary for the agent, but I hesitate whether we should now save data that is useful for external scripts/evals. I think usually we've been using the State pickle instead.

I agree we shouldn't have loads of extra data saved in the event stream. The `Condenser` base class already implements a metadata mechanism that lets the implementing classes call `self.add_metadata` and have the arguments registered in the `State` object.

The only bit of info being added here that isn't crucial is the `considered_event_ids` field. I included it for two reasons:
1. It fully captures the semantics of the condensation event -- even if the implementation of a condenser changes, we can use that extra information to unambiguously reconstruct histories from old versions.
2. It acts as a "checksum" in case the event stream gets messed up.

1 is definitely a personal preference, and 2 is probably an imagined issue. LMK and I can remove the field.

> * some logic could maybe be moved to conversation_memory, in theory it's its job to figure out the events in history and make up from them the context for the LLM. Such refactoring may help to make the code a bit more obvious here.

I'll have to think about that a bit, it'd be good to avoid complex back-and-forth between the conversation memory and the condenser. Agreed that there's plenty of room to refactor here (and I expect I'll be doing some very shortly).

enyst commented:
I've been thinking about this kind of case we have here, when we did Recall. The problem is familiar to you: we have a case of discriminated union, and no much underlying support for it in the current codebase. Like here, there we had RecallObservations, with two subtypes, `workspace_context` and `knowledge`, and a few more upcoming, e.g. `task` or `library_docs`.

I played with potential solutions, and the one that fit best IMHO was a class hierarchy, and the serialization done manually with a discriminator enum field (serialized as Recall, not as subtypes). If or when we migrate events from dataclasses to pydantic, it would've been probably painless on this aspect.

However, that kind of solution hit a problem: the event stream is flat. Mostly flat. Some composition is used, but even that not a lot, and some of it was due to external complexity I think. The similar case was FileRead IIRC. As you know, eventually Recall is flattened: a single observation class, with a `recall_type`, and right from the start, when created/deserialized, only one set of attrs is chosen. 
(that is, the subtypes remain unnamed/implicit) 🤔

I'm sorry for writing on the run, but I want to submit it for you to consider, because I think maybe such approach could be suited here too?

neubig commented:
This is a good discussion! Just a quick note on this:

> we should save in the stream data necessary for the agent, but I hesitate whether we should now save data that is useful for external scripts/evals

I think the event stream typically should hold information that's necessary for the agent, or also the human communicating with the agent. "considered_event_ids" seems like it's actually something that could be very useful for the human to trace what sort of information was considered in the summary (we could even surface it in the UI somehow), so I would vote in favor of keeping this as-is.

(separately, practically we'd like to finish up these condenser changes soon so we can turn condensation on by default soon, so whichever way we go it'd be good to make a timely decision!)

enyst commented:
> > we should save in the stream data necessary for the agent, but I hesitate whether we should now save data that is useful for external scripts/evals
> 
> I think the event stream typically should hold information that's necessary for the agent, or also the human communicating with the agent. "considered_event_ids" seems like it's actually something that could be very useful for the human to trace what sort of information was considered in the summary (we could even surface it in the UI somehow), so I would vote in favor of keeping this as-is.

I hear you, my point here is just that we already have it, from a simpler structure. The events that were summarized are `history[start_id..end_id]` or `history[keep_first...end_id]` I think.

We don't exactly _need_ to save the intermediary ids too, for summary.

> (separately, practically we'd like to finish up these condenser changes soon so we can turn condensation on by default soon, so whichever way we go it'd be good to make a timely decision!)

I think condenser is on by default already, isn't it?

With you on figuring it out sooner either way. I actually started working on this problem in https://github.com/All-Hands-AI/OpenHands/pull/7311 , then Calvin started too, from the other "end", then had to do the same thing 7311 did, so there was some duplicate work, which probably doesn't help a lot with timing, but it's also hard to avoid when we need a complete implementation so that we can check that it solves the bug it meant to solve.

I'm giving it a try on a test event stream.

enyst commented:
> The only bit of info being added here that isn't crucial is the `considered_event_ids` field (and maybe the `condenser_cls` field, but that'll be used when I update the agent controller's context truncation).

There is also `forgotten_ids`, or am I missing something? As far as I can tell, and please correct me if wrong, for all except LLM_ATTENTION, this field holds a list that we can, alternatively, compute from much less information (start, end; a slice of history).

I can see why we could also do this one for all, since we have to do it for one. I'm not fully sure though, with event streams of 3k+ events, that's a lot of info to send via the websocket, even, every step, and of course to save and to parse? 🤔 

> > * some logic could maybe be moved to conversation_memory, in theory it's its job to figure out the events in history and make up from them the context for the LLM. Such refactoring may help to make the code a bit more obvious here.
> 
> I'll have to think about that a bit, it'd be good to avoid complex back-and-forth between the conversation memory and the condenser. Agreed that there's plenty of room to refactor here (and I expect I'll be doing some very shortly).

I think one way to look at it is that the condenser is included in the "responsibility" of conversation_memory in a sense: not how it's doing things, that's mostly its thing, but what does the history sent to the LLM ends up having. That interface between the condenser and the agent could be actually an interface between condenser and conversation_memory (e.g. just include the call to condenser in `process_events` to begin with, it may help us think through the next steps?) 

Maybe then we see the next step could be that we don't need to return `events` from the condensers, we only need `AgentCondensationAction | None` ?

csmith49 commented:
Lots of good feedback here, thanks you two. I'll try and address the points made, but TL;DR I'd like to get this merged and resolve those concerns in future PRs if possible.

> I think condenser is on by default already, isn't it?

Not for the cloud version hosted by All Hands, there's a different settings chain and we kept it off until the context-forgetting issue was resolved.

> > The only bit of info being added here that isn't crucial is the `considered_event_ids` field (and maybe the `condenser_cls` field, but that'll be used when I update the agent controller's context truncation).
> 
> There is also `forgotten_ids`, or am I missing something? As far as I can tell, and please correct me if wrong, for all except LLM_ATTENTION, this field holds a list that we can, alternatively, compute from much less information (start, end; a slice of history).

Okay, so there's two issues here:

First, some fields (`considered_event_ids`) are not necessary for describing _how_ the condensation event modifies the history. But it's useful and human-parseable, and I agree with Graham it's worth keeping around.

Second, some fields (`considered_event_ids` and `forgotten_event_ids`) are not as compact as they could be. It sounds like this is a big concern for you @enyst, considering your proposal in #7526, so maybe I can address some of that here.

> I'd love your thoughts on https://github.com/All-Hands-AI/OpenHands/pull/7526 , as minimal (-ish) alternative. Could it have, or maybe it doesn't have all we actually need?

I actually tried a similar approach when working on this PR, but I found myself re-inventing the `pydantic` wheel to get out of the serialization slog. My concerns aren't representational, they're to do with code complexity and control-flow: without `pydantic` capabilities to do per-strategy field validation, history reconstruction gets pretty brittle and difficult to recover from.

One big benefit of a uniform condensation event is that history reconstruction is _uniform_. There's no delegation to condensers or a big switch statement over the condenser strategy (technically we have the former now but the implementations are _nearly_ the same and can be made the same without issue), both of which are sources of complexity hindering your suggested refactor re: condensers and conversation manager straightforward.

The single event approach won't be as space-efficient as some kind of union, but it really helps to keep code complexity down.

> I'm not fully sure though, with event streams of 3k+ events, that's a lot of info to send via the websocket, even, every step, and of course to save and to parse? 🤔

I don't know how the numbers shake out here tbh, but I expect the space-inefficient condensation events are a minor contributor to event stream bloat. With the default condenser, a condensation happens once every ~20 steps and forgets ~20 events. That means the inefficient condensation event (with a list of IDs instead of a start/end ID) stores an extra 18 integers. 

Right now I think it makes sense to err on the side of simpler code and worry about the optimizations later, especially if we're still hoping to refactor the conversation manager into the mix. If that's a deal-breaker for you I'd be happy to address the concerns in a follow-up PR (i.e., #7526).

enyst commented:
Sorry for the quick notes, I just want to point out I don't fully understand this, and maybe it's the source of the different impact we feel it has:

> With the default condenser, a condensation happens once every ~20 steps and forgets ~20 events. That means the inefficient condensation event (with a list of IDs instead of a start/end ID) stores an extra 18 integers.

Yes, the first time it happens. In a 2,000 event stream, though, that means plenty of AgentCondensationActions, each storing lists. It seems like the best we could hope for if we store lists, is that we only store an id at most once, which means AgentCondensationActions together store ~1,980, at the minimum, plus some overlap (not sure), plus the `considered` list.

enyst commented on openhands/memory/condenser/impl/amortized_forgetting_condenser.py:
```diff
@@ -32,18 +37,44 @@ def __init__(self, max_size: int = 100, keep_first: int = 0):
 
+        for event in events:
+            if isinstance(event, CondensationAction):
+                forgotten_event_ids.extend(event.forgotten_event_ids)
```
If I see this right, we use the `View` to compute the next `Event` attributes?

It's a little surprising, codewise, as it feels like we should be able to detach something that is a `view` from its underlying `model`, but here we actually use the View as the previous model, or do I misunderstand?

enyst commented on openhands/memory/condenser/impl/amortized_forgetting_condenser.py:
```diff
@@ -32,18 +37,44 @@ def __init__(self, max_size: int = 100, keep_first: int = 0):
+
+        event = CondensationAction(
+            forgotten_event_ids=[
+                event.id for event in view if event not in events_to_keep
```
Ah, not there. Maybe I misunderstood how it works. Here: if I see this right, we use the `View` to compute the next `Event` attributes?

It's a little surprising, codewise, as it feels like we should be able to detach something that is a `view` from its underlying `model`, but here we actually use the View as the previous model, or do I misunderstand?

csmith49 replied:
The model that the view is detached from is the history. The idea is that everything in the agent loop that would normally be looking at a history is instead looking at a `View`, which acts just like a list of events with some extra type safety.

Certain condensation actions (like summarizing) don't act over the entire history, just the subset that the agent loop would be seeing. So the control flow is this:

1. The condenser examines the history and applies any detected condensation events to produce a view.
2. The condenser determines if the view is too large and should be condensed.
3. If so, the condenser produces a new event that says how the _next_ view should be computed from the previous view.

The agent loop either gets the view or the condensation event. In the former case everything proceeds like normal, in the latter the agent can return the event and on the next step the condenser will produce the newest `View`.

csmith49 replied:
Although it's worth noting that the current condensation events are structured in such a way that computing the next view is really an action over the history and not over the previous view.

My goal was to make the context truncation that happens in the agent controller _also_ generate these events, and to move the `history -> View` calculation to something `Condenser` agnostic.

enyst replied:
You're right, we should consider the controller truncation bugged on this matter, and we're in full agreement that `history` should stay as a static list, everyone on the path doing what they need on it without modifying it.

(maybe the quick way to fix the controller might be to save `truncation_id` in the event rather than state, and never interpret it - so _not_ reload a truncated history in the controller - instead let `conversation_memory` interpret it and continue processing only starting from it or whatever fit. We may need to think how that works recursively.)

Back to the point here: the code is making a view first, then from it a `CondensationAction`. That's not a big issue, I'm just saying it's strange in the sense that the code feels like a view contribute directly at deciding underlying events... 🤔 

> how the next view should be computed from the _previous_ view.

Maybe that's what I don't fully get here. Why from the previous view, as opposed to from `history`?

If we have, e.g. for summary, the _range_ of events that were summarized (`start_id..end_id`), we don't seem to need to know that there was a previous slice of history that had some other range? Maybe we can see the next view as: `keep_first` + `summary event` + `tail`, or am I missing something? Depending on how the summary is done (includes previous summaries)

csmith49 replied:
> (maybe the quick way to fix the controller might be to save truncation_id in the event rather than state, and never interpret it - so not reload a truncated history in the controller - instead let conversation_memory interpret it and continue processing only starting from it or whatever fit. We may need to think how that works recursively.)

I've got a working prototype that uses the condensation event to handle the truncation in the agent, it's pretty minor changes from this PR as it stands. Currently the `conversation_memory` happens downstream of the condenser, though, which needs to know what the currently truncated history is to do the things it needs to do.

csmith49 replied:
> Why from the previous view, as opposed to from history?

The best performing condensers are the rolling condensers. They don't condense the history, they condense the previous condensation plus any new events. This lets us keep the considered history small while preserving cache hits and performance.

> If we have, e.g. for summary, the range of events that were summarized (start_id..end_id), we don't seem to need to know that there was a previous slice of history that had some other range? Maybe we can see the next view as: keep_first + summary event + tail, or am I missing something? Depending on how the summary is done (includes previous summaries)

If you have the range of events that were summarized, plus the summary, plus the full history, you could construct the new summary and the appropriate condensation event. All that is needed to find the critical info: the events that happened since the last summary and the start of the tail.

But all that calculation is already done to produce the view, which is needed for anything downstream anyways. So we condense the view instead of considering the entire history (most of which we'll throw away).

enyst replied:
That's a good point! Yes, it feels totally understandable... I feel like maybe I started thinking about this with the stream, not with inner details of condensing.

> All that is needed to find the critical info: the events that happened since the last summary and the start of the tail.

This seems to be, from `end_id` to `tail`.

> But all that calculation is already done to produce the view, which is needed for anything downstream anyways. So we condense the view instead of considering the entire history (most of which we'll throw away).

Right. Maybe it shouldn't be needed downstream this way. 🤔 After all, what we do is limit downstream to another `history` called `events`. The alternative here seems to be that we could maybe let the part of the code doing this anyway (interpreting `history` and figuring out what stays in and what goes out of context) do just that.

What I wonder is if we can easily rebuild `forgotten` and `considered` from a simple `Action` (or the previous condense actions), or we can't. Forgotten should be straightforward if we stored the start of the full forgotten array... `considered`... is considered the difference between `history` and `forgotten`?
- some code for thought: https://github.com/All-Hands-AI/OpenHands/pull/7534

csmith49 replied:
The view doesn't have to be computed by the condenser. But I worry about having condenser-specific actions and having the view computed elsewhere. You end up with this pattern of adding new condensers where you have to:

1. Implement the condenser interface.
2. Extend the polymorphic event to include any information specific to the condenser, without breaking any of the existing fields or serialization methods.
3. Extend the view computation to reference any condenser-specific fields so that it plays nicely with the others.

> What I wonder is if we can easily rebuild forgotten and considered from a simple Action (or the previous condense actions), or we can't. Forgotten should be straightforward if we stored the start of the full forgotten array... considered... is considered the difference between history and forgotten?

I'm not sure I understand. The code you linked is pretty much identical to what happens when the condenser rebuilds a view, so I think the answer is yes?

enyst replied:
> 1. Implement the condenser interface.
> 2. Extend the polymorphic event to include any information specific to the condenser, without breaking any of the existing fields or serialization methods.

This sounds right, and it's what happens with FileRead and Recall I think.

> Extend the view computation to reference any condenser-specific fields so that it plays nicely with the others.

We will need interpret the new event so that the remaining events to be included in context are computed. I'm not sure that we need to support rebuilding a context based on a `history` with many types of condenser in it. Is that a matter of concern?

> The code you linked is pretty much identical to what happens when the condenser rebuilds a view, so I think the answer is yes?

Ah, I was just trying to get the code closer to this PR, from the other side, from a simpler `AgentCondensationAction`. To see what it needs, so that if we try to do something like 7526, it doesn't imply any deep change of this PR.

csmith49 replied:
> I'm not sure that we need to support rebuilding a context based on a history with many types of condenser in it. Is that a matter of concern?

I _really_ hope not.

enyst commented:
> I'd classify it more as "debug" information than eval information. I can move it to the `State` but I also don't know how to tell a user with a messed up session how they should send me the associated `State` object like I do with the trajectory.

I see, I think maybe we can make sure that the flow between agent, with full `state.history`, and the creation of a condensation action is both well tested, and have logs. The trajectory will never contain everything in logs, after all. I can see that more information than the trajectory may be needed, preferably debug logs.

Also, if we have the event stream, maybe we're able to [replay](https://github.com/All-Hands-AI/OpenHands/issues/6049) it.

enyst commented on openhands/memory/condenser/condenser.py:
```diff
@@ -32,17 +33,54 @@ def get_condensation_metadata(state: State) -> list[dict[str, Any]]:
+class Condensation(BaseModel):
+    """Produced by a condenser to indicate the history has been condensed."""
+
+    action: CondensationAction
```
Nit (related to the simplification of the event stream that I was proposing) could you please tell, do we need a `Condensation` class, or could we use the `action`?

csmith49 commented:
@enyst I think I've addressed your concerns about the condensation event, please take a look and let me know. It sounds like that's the only thing that can't be resolved in future PRs?

enyst commented:
> EDIT TO ADD: I'm really trying to avoid flat representations of polymorphic events like we see in #7526. Very easy to do with `pydantic` but with the current system I keep running into all kinds of corner cases. 

I thought there are reasons in favor of it, such as:
- it's consistent, unfortunately (File action, Recall action)
- the fact that it can be done with pydantic is a plus today IMHO (because I do think we'll migrate at least in part, so in some sense, it's just the dataclass view on what will be; the serialized form would be the same I think)

The other options seem to be:
- a single event for all, as in... usable for all no matter what it contains? But...  can that be the case?
- many events (hopefully we don't consider this, I briefly did it for Recall and it didn't get buy-in)

> This approach has a single event (with an unavoidable instance of field polymorphism) that can be used by all existing condensers. There's a big assumption about event IDs in the range calculations though -- can you confirm that's an okay assumption to make?

I'm not sure I understand, sorry, could you elaborate? 

If it was created with a condenser type, personally I wouldn't expect it to be read by another condenser type. In other words, I wouldn't expect to switch condenser type and still be able to map every condense action to an actual state of the new one.

enyst commented on openhands/agenthub/codeact_agent/codeact_agent.py:
```diff
@@ -113,8 +116,23 @@ def step(self, state: State) -> Action:
+        # to the conversation manager for processing, but if we get a condensation
+        # event we'll just return that instead of an action. The controller will
+        # immediately ask the agent to step again with the new view.
+        condensed_history: list[Event] = []
```
Sorry, just to understand, is there a reason why we changed the order of processing?

csmith49 replied:
This condensation is pulled out of `self._get_messages` because it can return an `Action` that should be returned and I didn't want to mess with the return types of existing functions. When there's no condensation this does not change the order of processing.

When there _is_ a condensation this has the effect of adding the action to the event stream, which means any future views computed will be done wrt that condensation.

This is similar control-flow to the current context truncation in the agent controller, where some state is changed and we want to "try again" by emitting an event to trigger a new step (only here that event contains critical information for view calculations).

enyst replied:
Ah, of course. Not a problem, just a remark: this makes it a little more difficult to see that the condenser doesn't _need_ to return events. It needs to return a single event if any.

neubig replied:
Sounds good, resolving this conversation to make it easier to track the things we still need to resolve.

enyst commented on openhands/events/action/agent.py:
```diff
@@ -111,3 +111,87 @@ def __str__(self) -> str:
+    1. By providing a list of event IDs.
+    2. By providing the start and end IDs of a range of events.
+
+    In the second case, we assume that event IDs are monotonically increasing, and that _all_ events between the start and end IDs are to be forgotten.
```
Yes, I also noticed again while rebuilding lists that this is completely reasonable, but for the condense actions themselves. They will be in the stream/history interweaved with the last big list to be forgotten.

This isn't just an assumption, it's by definition. We can define the start id and end id to be the range between which most condensers (the exception being attention) have their forgotten events in that step.

In fact, whether we need the other condense actions is a question for the condenser implementation. Outside the condensers, if we want to send them to the agent LLM, we can pick them up. They're in the stream, so they're in history, so we don't, strictly speaking, need the condensers to _also_ provide those, we can pick them up if we want.

csmith49 replied:
Just to confirm so we can resolve this comment -- _all_ event IDs are ordered monotonically increasing as they get added to the event stream? So if `event_i` is added to the event stream, then `event_j` is added, it will always be the case that `event_i.id < event_j.id`?

enyst replied:
Yes. The stream itself assigns ids in ascending order, in the `add_event` method, which is the only way to do it, and the `_id` field is otherwise never set.

csmith49 commented:
> > EDIT TO ADD: I'm really trying to avoid flat representations of polymorphic events like we see in #7526. Very easy to do with `pydantic` but with the current system I keep running into all kinds of corner cases.
> 
> I thought there are reasons in favor of it, such as:
> ...

Sorry, I don't mean to imply it's a bad design pattern in general. In this PR I just kept running into issues where I'd have a huge set of fields and a single discriminator enum that would control which combinations of fields should be enabled and how they should be interpreted.

I can add validators in `__post_init__` to check these configurations, but I can't make the data class frozen or use that validation to enforce type guarantees so there were always a lot of unnecessary assertions when trying to get access to those combinations of fields.

> * a single event for all, as in... usable for all no matter what it contains? But...  can that be the case?

That's the approach I tried to go with. The current implementation has that working, but I couldn't avoid all instances of that flat polymorphism pattern without just relying on lists of integers for field identifiers.

> > There's a big assumption about event IDs in the range calculations though -- can you confirm that's an okay assumption to make?
> 
> I'm not sure I understand, sorry, could you elaborate?

The assumption is that events have monotonically increasing (wrt their creation time) IDs. That way I can define start/end ranges based on IDs and -- since the forgotten events act as a filter over the history -- I won't miss any events or forget one that I don't want to.

enyst commented on openhands/events/action/agent.py:
```diff
@@ -111,3 +111,87 @@ def __str__(self) -> str:
+    """An optional summary of the events being forgotten."""
+
+    summary_offset: int | None = None
+    """An optional offset to the start of the resulting view indicating where the summary should be inserted."""
```
Sorry to ask on the fly, while reading, but why do we need this? It seems like we know where a condense action should be inserted: in order, old to new, the order in which they are in the stream/history, and, starting just after keep_first?

e.g. keep_first, oldest condense, ... newest condense, tail ?

csmith49 replied:
Right, just after `keep_first`. But only the condenser knows that parameter, so we either rely on condensers to compute the views (which I'm trying to avoid with the uniform `CondensationAction`) or we provide a summary offset that corresponds to `keep_first` so the view is totally reconstructable from just the event stream.

enyst replied:
Actually, we know it elsewhere... It's the `start_id` of the oldest `condense action` in the stream / history?

csmith49 replied:
I believe that's correct? It's an assumption we're probably best avoiding during reconstruction, though, unless we're documenting it very heavily.

enyst replied:
I wonder why would we avoid it? It seems to me it's not really assumption, it's by definition.

If we set `start_id` and `end_id` to mark the forgotten events at every step, it has to work that way. How else?

enyst commented:
> Sorry, I don't mean to imply it's a bad design pattern in general. In this PR I just kept running into issues where I'd have a huge set of fields and a single discriminator enum that would control which combinations of fields should be enabled and how they should be interpreted.

I see... was that an artefact of the combinations of fields in the work in progress or does it have to be the case always? I ask because I'm not sure the combinations are many... with the range, I mean.

> The assumption is that events have monotonically increasing (wrt their creation time) IDs. That way I can define start/end ranges based on IDs and -- since the forgotten events act as a filter over the history -- I won't miss any events or forget one that I don't want to.

Yes, this is reasonable. It is definitely the case in the stream, and it is the case in history save for the truncation in action. I can look into stopping that, but if this PR works as expected, that case should be extremely rare with Claude and other LLMs with decent-ish windows.

enyst commented on openhands/events/action/agent.py:
```diff
@@ -111,3 +111,87 @@ def __str__(self) -> str:
+    forgotten_event_ids: list[int] | None = None
+    """The IDs of the events that are being forgotten (removed from the `View` given to the LLM)."""
+
+    forgotten_events_start_id: int | None = None
```
Nit: this is accurate, and it's cool if it helps during refactoring, it's always the same though, so maybe we don't need "forgotten_events" in its name. Ditto below.

csmith49 replied:
If all we had to store were events, I think I'd agree, but with the summary field(s) also present I don't think `start_id` and `end_id` are descriptive enough.

enyst replied:
OK!

enyst commented on openhands/events/action/agent.py:
```diff
@@ -111,3 +111,87 @@ def __str__(self) -> str:
+        assert self.forgotten_events_start_id is not None
+        assert self.forgotten_events_end_id is not None
+        return list(
+            range(self.forgotten_events_start_id, self.forgotten_events_end_id + 1)
```
Ah, I see, you see this as being part of the event. 🤔 Trying to do it in the `Event`... should work, but if we need this `forgotten` property minus the previous `condense actions`, then I guess it won't work as such.

The extra properties may also mean we need to fix stuff in the serialization code for them, I think, though I could be wrong.

csmith49 replied:
> Ah, I see, you see this as being part of the event. 🤔 Trying to do it in the `Event`... should work, but if we need this `forgotten` property minus the previous `condense actions`, then I guess it won't work as such.

The best way I could find to handle the field (I guess it's normally called row) polymorphism. We only need that `forgotten_*` info to reconstruct the context, and that's done in an environment where the `CondensationAction` events are excluded anyways.

> The extra properties may also mean we need to fix stuff in the serialization code for them, I think, though I could be wrong.

They shouldn't -- looks like serialization rn is handled via `dataclasses.asdict`, which won't serialize functions with the property decorator.

enyst replied:
I see, thank you, OK! That's cool. It seems a better way to keep compatibility with condensers implementation than an external function/method like the one I made.

Sorry, the point of making it was that I would have loved to avoid more work on this PR, because I did and do believe that it's actually compatible without significant changes.

csmith49 commented:
> I see... was that an artefact of the combinations of fields in the work in progress or does it have to be the case always? I ask because I'm not sure the combinations are many... with the range, I mean.

It was definitely an issue with an earlier approach that captured more info. The options are not numerous here, as you can see in `CondensationAction`.

enyst commented on openhands/memory/condenser/impl/llm_summarizing_condenser.py:
```diff
@@ -32,26 +36,22 @@ def __init__(self, llm: LLM, max_size: int = 100, keep_first: int = 1):
-            events[self.keep_first]
-            if isinstance(events[self.keep_first], AgentCondensationObservation)
+            view[self.keep_first]
+            if isinstance(view[self.keep_first], AgentCondensationObservation)
```
```suggestion
            if isinstance(view[self.keep_first], CondensationAction)
```
It's an Action, in this PR, right?

csmith49 replied:
Different events for different uses. The action tells us to condense something in the history, the observation is the summary we send to the LLM.

csmith49 replied:
> The action has the summary, though, of course, it needs to have it.

The action has the summary, yes, but we don't have access to the actions while computing the next condensation, just the view that was produced. We could also pass the list of condensation actions in as a field of the `View` data class but I don't know that it helps clean things up much.

> Not for this PR IMHO, one alternative could be to remove the use of the obs in condensers and keep it only for the emergency truncation - as you were suggesting before, to separate the two.

The observation is what actually gets passed to the LLM, though. I'm happy to rename the observation to make things more clear (we inherited the observation name from the OG condenser implementation), but they are distinct events and shouldn't be made the same.

enyst replied:
I see, thank you. You may be right! 🤔If it keeps things manageable in condensers, maybe it's not an issue... We could also do something else with the truncation obs.

enyst commented:
Sorry, I tried to test that it works as expected, when restoring a session:
- made on main branch
- made on the PR branch

It doesn't seem to work at the moment because it won't recognize the old condense action? It creates actions, but it recognizes obs.

enyst commented:
I restored a ~450 events session made from the `main` branch. It's the one I was mentioning before, it's at the edge of a natural ContextWindowExceededLimit.

![image](https://github.com/user-attachments/assets/9ee3dcd2-2fbe-4d97-bac8-ebf92bcc6fb3)

It doesn't manage it, it falls back to the controller truncation, and then condensation happens at half (event_id=231). I think this is because condensers don't have error handling, and it gets the exception in the condenser LLM call.
- the "trimming prompt" from truncation is twice because I stopped it from the debugger, so it didn't save state. This looks normal to me.
- when it converts into `Messages` it sends all 3 condense observations.

csmith49 commented:
> Sorry, I tried to test that it works as expected, when restoring a session...made on main branch

I don't think we should expect that to work in this case. The old sessions won't have the `CondensationAction` events threaded through the history, so we'll try just one big condensation on the first agent step.

> It doesn't manage it, it falls back to the controller truncation, and then condensation happens at half (event_id=231). I think this is because condensers don't have error handling, and it gets the exception in the condenser LLM call.

That's also expected here, and how it currently works on `main`. I definitely want to improve the back-and-forth between the two, and I think `CondensationAction` events are the current shortest path to doing so.

csmith49 commented:
> Edited to add: Ah, we also create an obs with the same summary. Not sure if it's necessary but that obs won't be saved. Maybe that doesn't matter?

That doesn't matter. We can't save that observation in the right spot in the history without violating the temporal ordering of events, so the `CondensationAction` carries all necessary information for the `View` reconstruction to put it in the right spot before handing the view off to the LLM.

csmith49 commented on openhands/events/action/agent.py:
```diff
@@ -111,3 +111,87 @@ def __str__(self) -> str:
+class CondensationAction(Action):
+    """This action indicates a condensation of the conversation history is happening.
+
+    There are two ways to specify the events to be forgotten:
+    1. By providing a list of event IDs.
+    2. By providing the start and end IDs of a range of events.
```
The first path is a functional super-set of the second, yes. Engel expressed concerns about the size of the messages so I added the suggested range representation as an optimization. I don't have numbers on how much it saves but I think the overall memory impact is minor either way (maybe ~0.1% of the data in the whole trajectory?).

enyst replied:
Yes, I'm sorry, it may be just me and my inefficiency hate: they're consecutive, so unnecessary and they would be saved/read in every condensation action. 🥹 They would also be sent via websocket.

Our usual case is the summarizer, where they're consecutive. In fact, all of them, except for `llm_attention`, specify, at most, a range: from `start` to `end`.

IMHO a way to represent these would be with polymorphism and discriminated union. But the event stream is "flat". When we represented this kind of case before, we made a flattened representation with an enum as discriminator for serialization and some validation, and an umbrella class for the necessary fields. Calvin doesn't like that, quite understandably - and here they would be more than two. 😅

This is an umbrella class, but with just minimal options, which is two. It's a middle-ground I'm happy with, but maybe it's bad?

neubig replied:
Yeah, personally I think code complexity is worse than small amounts of inefficiency. But it's not a hill that I'll die on :)

enyst commented:
I have failed to test on the PR branch (for reasons unrelated to the PR, I'm sure), I can give it another try just now. IMHO otherwise the PR looks great, and I just wanted to see what it does on session restore.
